### PR TITLE
Automated Docker image upload

### PR DESCRIPTION
**Pull request by lucafavatellawork**
TODO:
- [x] Test actual upload to Docker Hub of branch - not master.
  - See [Travis CI job that uploaded image](https://travis-ci.com/opensparkl/sse_core/jobs/294922945), [Docker Hub tags](https://hub.docker.com/r/sparkl/sse/tags?name=docker-image-upload&page=1) and [Docker Hub image by hash](https://hub.docker.com/layers/sparkl/sse/docker-image-upload/images/sha256-a011dab3c92a2eef82ea45c71bb192967e950ffbf72a470ed5e2b35c8d737089?context=explore).
- [x] Revert branch to master in `.travis.yml`.
- [x] Revert any test changes in `.travis.yml`.
- [x] Ensure repo settings have secrets for Docker Hub in master.
- [ ] Ensure repo settings have no secrets for Docker Hub apart from master.

### CODE REVIEW

**Comment by lucafavatellawork** [comment]
I did not test that the CI actually triggers at the PR merge, though I suggest we figure out at the actual merge.

File: .travis.yml [hunk 1]
```diff
@@ -8,16 +8,36 @@ aliases:
...
-    - if: branch IN (master) AND repo = "opensparkl/sse_core" AND type NOT IN (pull_request)
+    - stage: deploy
       language: minimal
       git:
         depth: false
       services:
         - docker
       script:
         - make -C docker docker
+        - make -C docker latest_docker
```
Comment (miklosduma) [inline .travis.yml]: If I understand correctly, we do the following:
1. Do a docker build -> sparkl/sse
2. Run latest_docker to tag it sparkl/sse:latest
3. Re-tag it as sparkl/sse:master
4. Remove the sparkl/sse:latest tag
5. Push sparkl/sse:master

Wouldn't it be easier to leave out the latest_docker step and immediately tag the image with the branch name?

File: .travis.yml [hunk 2]
```diff
@@ -8,16 +8,36 @@ aliases:
...
+        - docker tag sparkl/sse:latest sparkl/sse:"${TRAVIS_BRANCH:?}"
+        - docker rmi sparkl/sse:latest
+      before_deploy:
+        - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+      deploy:
+        on:
+          all_branches: true  # Job is branch-agnostic.
```
Comment (miklosduma) [inline .travis.yml]: I see, you do the filtering already for push/master and all that

File: .travis.yml (hunk 1 above)
Comment (jacobythwaites) [reply .travis.yml]: I would agree with that, no need to tag it `latest`

Comment (lucafavatellawork) [reply .travis.yml]: That would require changing the Makefile, potentially with introduction of a new env var to pass the Makefile. That seems more complex and unneeded hence it looked good enough.

The whole point of having a well-known tag `latest` locally is in order not to try to infer dynamically the git-version-hash that is used for tagging the raw image.

File: .travis.yml (hunk 2 above)
Comment (lucafavatellawork) [reply .travis.yml]: Yes, I do the filtering at stage level.

File: .travis.yml (hunk 1 above)
Comment (jacobythwaites) [reply .travis.yml]: Locally using latest_docker is merely a nice trick to locally get hold of the generated image using a static name. The :latest tag never goes outside the build machine. So leave it, it's fine!

**Comment by lucafavatellawork** [comment]
> I did not test that the CI actually triggers at the PR merge, though I suggest we figure out at the actual merge.

It does trigger.

**Pull request merged**